File: aep-lib-go/pkg/api/openapi.go
package api

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/aep-dev/aep-lib-go/pkg/cases"
	"github.com/aep-dev/aep-lib-go/pkg/constants"
	"github.com/aep-dev/aep-lib-go/pkg/openapi"
)

const (
	AEP_OPERATION_REF = "https://aep.dev/json-schema/type/operation.json"
)

func (api *API) ConvertToOpenAPIBytes() ([]byte, error) {
	openAPI, err := ConvertToOpenAPI(api)
	if err != nil {
		return nil, err
	}

	jsonData, err := json.MarshalIndent(openAPI, "", "  ")
	if err != nil {
		return nil, err
	}
	return jsonData, nil
}

func ConvertToOpenAPI(api *API) (*openapi.OpenAPI, error) {
	paths := map[string]*openapi.PathItem{}
	components := openapi.Components{
		Schemas: map[string]openapi.Schema{},
	}
	for _, r := range api.Resources {
		// Ensure r.Schema is not nil before dereferencing
		if r.Schema == nil {
			return nil, fmt.Errorf("schema for resource %s is nil", r.Singular)
		}
		d := r.Schema
		removeXAEPFieldNumber(d)
		// if it is a resource, add paths
		collection, parentPWPS := generateParentPatternsWithParams(r)
		// Ensure parentPWPS is not nil before dereferencing
		if parentPWPS == nil {
			return nil, fmt.Errorf("parent patterns for resource %s are nil", r.Singular)
		}
		// add an empty PathWithParam, if there are no parents.
		// This will add paths for the simple resource case.
		if len(*parentPWPS) == 0 {
			*parentPWPS = append(*parentPWPS, PathWithParams{
				Pattern: "", Params: []openapi.Parameter{},
			})
		}
		patterns := []string{}
		schemaRef := fmt.Sprintf("#/components/schemas/%v", r.Singular)
		resourceSchema := &openapi.Schema{
			Ref: schemaRef,
		}
		singular := r.Singular
		// Convert kebab-case singular to snake_case for path variables
		singularSnake := cases.KebabToSnakeCase(singular)
		// declare some commonly used objects, to be used later.
		bodyParam := openapi.RequestBody{
			Required: true,
			Content: map[string]openapi.MediaType{
				"application/json": {
					Schema: resourceSchema,
				},
			},
		}
		idParam := openapi.Parameter{
			In:       "path",
			Name:     singularSnake + "_id",
			Required: true,
			Schema: &openapi.Schema{
				Type: "string",
			},
		}
		resourceResponse := openapi.Response{
			Description: "Successful response",
			Content: map[string]openapi.MediaType{
				"application/json": {
					Schema: resourceSchema,
				},
			},
		}
		for _, pwp := range *parentPWPS {
			resourcePath := fmt.Sprintf("%s%s/{%s_id}", pwp.Pattern, collection, singularSnake)
			patterns = append(patterns, resourcePath[1:])
			if r.Methods.List != nil {
				listPath := fmt.Sprintf("%s%s", pwp.Pattern, collection)
				responseProperties := map[string]openapi.Schema{
					constants.FIELD_RESULTS_NAME: {
						Type:  "array",
						Items: resourceSchema,
					},
					constants.FIELD_NEXT_PAGE_TOKEN_NAME: {
						Type: "string",
					},
				}
				if r.Methods.List.HasUnreachableResources {
					responseProperties[constants.FIELD_UNREACHABLE_NAME] = openapi.Schema{
						Type: "array",
						Items: &openapi.Schema{
							Type: "string",
						},
					}
				}
				params := append(pwp.Params,
					openapi.Parameter{
						In:       "query",
						Name:     constants.FIELD_MAX_PAGE_SIZE_NAME,
						Required: false,
						Schema: &openapi.Schema{
							Type: "integer",
						},
					},
					openapi.Parameter{
						In:       "query",
						Name:     constants.FIELD_PAGE_TOKEN_NAME,
						Required: false,
						Schema: &openapi.Schema{
							Type: "string",
						},
					},
				)

				if r.Methods.List.SupportsSkip {
					params = append(params, openapi.Parameter{
						In:       "query",
						Name:     constants.FIELD_SKIP_NAME,
						Required: false,
						Schema: &openapi.Schema{
							Type: "integer",
						},
					})
				}
				if r.Methods.List.SupportsFilter {
					params = append(params, openapi.Parameter{
						In:       "query",
						Name:     constants.FIELD_FILTER_NAME,
						Required: false,
						Schema: &openapi.Schema{
							Type: "string",
						},
					})
				}
				methodInfo := openapi.Operation{
					OperationID: fmt.Sprintf("List%s", cases.SnakeToPascalCase(singularSnake)),
					Description: fmt.Sprintf("List method for %s", r.Singular),
					Parameters:  params,
					Responses: map[string]openapi.Response{
						"200": {
							Description: "Successful response",
							Content: map[string]openapi.MediaType{
								"application/json": {
									Schema: &openapi.Schema{
										Type:       "object",
										Properties: responseProperties,
									},
								},
							},
						},
					},
				}
				addMethodToPath(paths, listPath, "get", methodInfo)
			}
			if r.Methods.Create != nil {
				createPath := fmt.Sprintf("%s%s", pwp.Pattern, collection)
				params := pwp.Params
				if r.Methods.Create.SupportsUserSettableCreate {
					params = append(params, openapi.Parameter{
						In:       "query",
						Name:     "id",
						Required: false,
						Schema: &openapi.Schema{
							Type: "string",
						},
					})
				}
				methodInfo := openapi.Operation{
					OperationID: fmt.Sprintf("Create%s", cases.SnakeToPascalCase(singularSnake)),
					Description: fmt.Sprintf("Create method for %s", r.Singular),
					Parameters:  params,
					RequestBody: &bodyParam,
					Responses: map[string]openapi.Response{
						"200": resourceResponse,
					},
				}
				if r.Methods.Create.IsLongRunning {
					methodInfo.XAEPLongRunningOperation = &openapi.XAEPLongRunningOperation{
						Response: openapi.XAEPLongRunningOperationResponse{
							Schema: resourceSchema,
						},
					}
					methodInfo.Responses = map[string]openapi.Response{
						"200": {
							Description: "Long-running operation response",
							Content: map[string]openapi.MediaType{
								"application/json": {
									Schema: &openapi.Schema{
										Ref: AEP_OPERATION_REF,
									},
								},
							},
						},
					}
				}
				addMethodToPath(paths, createPath, "post", methodInfo)
			}
			if r.Methods.Get != nil {
				methodInfo := openapi.Operation{
					OperationID: fmt.Sprintf("Get%s", cases.SnakeToPascalCase(singularSnake)),
					Description: fmt.Sprintf("Get method for %s", r.Singular),
					Parameters:  append(pwp.Params, idParam),
					Responses: map[string]openapi.Response{
						"200": resourceResponse,
					},
				}
				addMethodToPath(paths, resourcePath, "get", methodInfo)
			}
			if r.Methods.Update != nil {
				methodInfo := openapi.Operation{
					OperationID: fmt.Sprintf("Update%s", cases.SnakeToPascalCase(singularSnake)),
					Description: fmt.Sprintf("Update method for %s", r.Singular),
					Parameters:  append(pwp.Params, idParam),
					RequestBody: &openapi.RequestBody{
						Required: true,
						Content: map[string]openapi.MediaType{
							"application/merge-patch+json": {
								Schema: &openapi.Schema{
									Ref: schemaRef,
								},
							},
						},
					},
					Responses: map[string]openapi.Response{
						"200": {
							Description: "Successful response",
							Content: map[string]openapi.MediaType{
								"application/json": {
									Schema: &openapi.Schema{
										Ref: schemaRef,
									},
								},
							},
						},
					},
				}
				if r.Methods.Update.IsLongRunning {
					methodInfo.XAEPLongRunningOperation = &openapi.XAEPLongRunningOperation{
						Response: openapi.XAEPLongRunningOperationResponse{
							Schema: resourceSchema,
						},
					}
					methodInfo.Responses = map[string]openapi.Response{
						"200": {
							Description: "Long-running operation response",
							Content: map[string]openapi.MediaType{
								"application/json": {
									Schema: &openapi.Schema{
										Ref: AEP_OPERATION_REF,
									},
								},
							},
						},
					}
				}
				addMethodToPath(paths, resourcePath, "patch", methodInfo)
			}
			if r.Methods.Delete != nil {
				responseSchema := &openapi.Schema{}
				params := append(pwp.Params, idParam)
				if len(r.Children) > 0 {
					params = append(params, openapi.Parameter{
						In:       "query",
						Name:     constants.FIELD_FORCE_NAME,
						Required: false,
						Schema: &openapi.Schema{
							Type: "boolean",
						},
					})
				}
				methodInfo := openapi.Operation{
					OperationID: fmt.Sprintf("Delete%s", cases.SnakeToPascalCase(singularSnake)),
					Description: fmt.Sprintf("Delete method for %s", r.Singular),
					Parameters:  params,
					Responses: map[string]openapi.Response{
						"204": {
							Description: "Successful response",
							Content: map[string]openapi.MediaType{
								"application/json": {
									Schema: responseSchema,
								},
							},
						},
					},
				}
				if r.Methods.Delete.IsLongRunning {
					methodInfo.XAEPLongRunningOperation = &openapi.XAEPLongRunningOperation{
						Response: openapi.XAEPLongRunningOperationResponse{
							Schema: responseSchema,
						},
					}
					methodInfo.Responses = map[string]openapi.Response{
						"200": {
							Description: "Long-running operation response",
							Content: map[string]openapi.MediaType{
								"application/json": {
									Schema: &openapi.Schema{
										Ref: AEP_OPERATION_REF,
									},
								},
							},
						},
					}
				}
				addMethodToPath(paths, resourcePath, "delete", methodInfo)
			}
			if r.Methods.Apply != nil {
				methodInfo := openapi.Operation{
					OperationID: fmt.Sprintf("Apply%s", cases.SnakeToPascalCase(singularSnake)),
					Description: fmt.Sprintf("Apply method for %s", r.Singular),
					Parameters:  append(pwp.Params, idParam),
					RequestBody: &bodyParam,
					Responses: map[string]openapi.Response{
						"200": resourceResponse,
					},
				}
				if r.Methods.Apply.IsLongRunning {
					methodInfo.XAEPLongRunningOperation = &openapi.XAEPLongRunningOperation{
						Response: openapi.XAEPLongRunningOperationResponse{
							Schema: resourceSchema,
						},
					}
					methodInfo.Responses = map[string]openapi.Response{
						"200": {
							Description: "Long-running operation response",
							Content: map[string]openapi.MediaType{
								"application/json": {
									Schema: &openapi.Schema{
										Ref: AEP_OPERATION_REF,
									},
								},
							},
						},
					}
				}
				addMethodToPath(paths, resourcePath, "put", methodInfo)
			}
			for _, custom := range r.CustomMethods {
				// Ensure custom.Response and custom.Request are not nil
				if custom.Response == nil {
					custom.Response = &openapi.Schema{
						Type: "object",
					}
				}
				if custom.Request == nil {
					custom.Request = &openapi.Schema{
						Type: "object",
					}
				}
				// Remove XAEPFieldNumber from custom method schemas
				removeXAEPFieldNumber(custom.Request)
				removeXAEPFieldNumber(custom.Response)
				methodType := "get"
				if custom.Method == "POST" {
					methodType = "post"
				}
				cmPath := fmt.Sprintf("%s:%s", resourcePath, custom.Name)
				methodInfo := openapi.Operation{
					OperationID: fmt.Sprintf(":%s%s", cases.SnakeToPascalCase(custom.Name), cases.SnakeToPascalCase(singularSnake)),
					Description: fmt.Sprintf("Custom method %s for %s", custom.Name, r.Singular),
					Parameters:  append(pwp.Params, idParam),
					Responses: map[string]openapi.Response{
						"200": {
							Description: "Successful response",
							Content: map[string]openapi.MediaType{
								"application/json": {
									Schema: custom.Response,
								},
							},
						},
					},
				}
				if custom.Method == "POST" {
					methodInfo.RequestBody = &openapi.RequestBody{
						Required: true,
						Content: map[string]openapi.MediaType{
							"application/json": {
								Schema: custom.Request,
							},
						},
					}
				}
				// Ensure the response schema for long-running operations is correctly set
				if custom.IsLongRunning {
					methodInfo.XAEPLongRunningOperation = &openapi.XAEPLongRunningOperation{
						Response: openapi.XAEPLongRunningOperationResponse{
							Schema: custom.Response,
						},
					}
					methodInfo.Responses = map[string]openapi.Response{
						"200": {
							Description: "Long-running operation response",
							Content: map[string]openapi.MediaType{
								"application/json": {
									Schema: &openapi.Schema{
										Ref: AEP_OPERATION_REF,
									},
								},
							},
						},
					}
				}
				addMethodToPath(paths, cmPath, methodType, methodInfo)
			}
		}
		d.XAEPResource = &openapi.XAEPResource{
			Singular: r.Singular,
			Plural:   r.Plural,
			Patterns: patterns,
			Parents:  r.Parents,
			Type:     fmt.Sprintf("%s/%s", api.Name, r.Singular),
		}
		components.Schemas[r.Singular] = *d
	}
	for k, v := range api.Schemas {
		// Create a copy of the schema to avoid modifying the original
		schemaCopy := *v
		removeXAEPFieldNumber(&schemaCopy)
		components.Schemas[k] = schemaCopy
	}

	contact := openapi.Contact{}
	if api.Contact != nil {
		contact = openapi.Contact{
			Name:  api.Contact.Name,
			Email: api.Contact.Email,
			URL:   api.Contact.URL,
		}
	}
	openAPI := &openapi.OpenAPI{
		OpenAPI: "3.1.0",
		Servers: []openapi.Server{
			{URL: api.ServerURL},
		},
		Info: openapi.Info{
			Title:       api.Name,
			Version:     "version not set",
			Description: "An API for " + api.Name,
			Contact:     contact,
		},
		Paths:      paths,
		Components: components,
	}
	return openAPI, nil
}

// PathWithParams passes an http path
// with the OpenAPI parameters it contains.
// helpful to bundle them both when iterating.
type PathWithParams struct {
	Pattern string
	Params  []openapi.Parameter
}

// generate the x-aep-patterns for the parent resources, along with the patterns
// they need. Return a tuple of the collection name for the resource, and the
// patterns.
//
// This is helpful when you're constructing methods on resources with a parent.
//
// There are two algorithms that are used:
//
// 1. if PatternElems are present, then those will be used. This helps
// handle the situation where the resource structs were retrieved from a parsed
// OpenAPI definition, where the plural of the parents aren't necessarily clear,
// or the pattern element naming may not completely match the resource names.
//
// 2. Otherwise, we'll use the parent resources, and generate the collection
// names. This works for the case where the resource hierarchy is generated from
// scratch. This Algorithm will result in the fully AEP-compliant collection
// names.
func generateParentPatternsWithParams(r *Resource) (string, *[]PathWithParams) {
	// case 1: pattern elems are present, so we use them.
	// TODO(yft): support multiple patterns
	if len(r.patternElems) > 0 {
		collection := fmt.Sprintf("/%s", r.patternElems[len(r.patternElems)-2])
		params := []openapi.Parameter{}
		for i := 0; i < len(r.patternElems)-2; i += 2 {
			pElem := r.patternElems[i+1]
			// Extract the parameter name without the _id suffix
			paramName := pElem[1 : len(pElem)-1]
			params = append(params, openapi.Parameter{
				In:       "path",
				Name:     paramName,
				Required: true,
				Schema: &openapi.Schema{
					Type: "string",
				},
			})
		}
		pattern := strings.Join(r.patternElems[0:len(r.patternElems)-2], "/")
		if pattern != "" {
			pattern = fmt.Sprintf("/%s", pattern)
		}
		return collection, &[]PathWithParams{
			{Pattern: pattern, Params: params},
		}
	}
	// case 2: no pattern elems, so we need to generate the collection names
	collection := fmt.Sprintf("/%s", CollectionName(r))
	pwps := []PathWithParams{}
	for _, parent := range r.ParentResources() {
		singular := parent.Singular
		// Convert kebab-case singular to snake_case for path variables
		singularSnakeParam := cases.KebabToSnakeCase(singular) + "_id"
		basePattern := fmt.Sprintf("/%s/{%s}", CollectionName(parent), singularSnakeParam)
		baseParam := openapi.Parameter{
			In:       "path",
			Name:     singularSnakeParam,
			Required: true,
			Schema: &openapi.Schema{
				Type: "string",
			},
			XAEPField: &openapi.XAEPField{
				ResourceReference: []string{singular},
			},
		}
		if len(parent.ParentResources()) == 0 {
			pwps = append(pwps, PathWithParams{
				Pattern: basePattern,
				Params:  []openapi.Parameter{baseParam},
			})
		} else {
			_, parentPWPS := generateParentPatternsWithParams(parent)
			for _, parentPWP := range *parentPWPS {
				params := append(parentPWP.Params, baseParam)
				pattern := fmt.Sprintf("%s%s", parentPWP.Pattern, basePattern)
				pwps = append(pwps, PathWithParams{Pattern: pattern, Params: params})
			}
		}
	}
	return collection, &pwps
}

func addMethodToPath(paths map[string]*openapi.PathItem, path, method string, methodInfo openapi.Operation) {
	methods, ok := paths[path]
	if !ok {
		methods = &openapi.PathItem{}
		paths[path] = methods
	}
	switch method {
	case "get":
		methods.Get = &methodInfo
	case "post":
		methods.Post = &methodInfo
	case "patch":
		methods.Patch = &methodInfo
	case "put":
		methods.Put = &methodInfo
	case "delete":
		methods.Delete = &methodInfo
	}
}

// removeXAEPFieldNumber removes the field_number from the schema.
// Currently field_number is not an official annotation in the AEPs,
// and adding it will confuse consumers.
func removeXAEPFieldNumber(schema *openapi.Schema) {
	if schema.XAEPField != nil {
		schema.XAEPField.FieldNumber = 0
		if schema.XAEPField.Behavior == nil && schema.XAEPField.ResourceReference == nil && schema.XAEPField.ResourceReferenceChildType == nil {
			schema.XAEPField = nil
		}
	}
	if schema.Properties != nil {
		for key, property := range schema.Properties {
			removeXAEPFieldNumber(&property)
			schema.Properties[key] = property
		}
	}
}

File: aep-lib-ts/src/api/openapi.ts
import {
  API,
  Contact,
  OpenAPI,
  PathItem as OpenAPIPathItem,
  PathWithParams,
  Resource,
  APISchema,
  Operation,
  RequestBody,
  Parameter,
  Response,
} from "./types.js";
import { collectionName } from "./resource.js";
import { kebabToPascalCase } from "../cases/cases.js";

export function convertToOpenAPI(api: API): OpenAPI {
  const paths: Record<string, OpenAPIPathItem> = {};
  const components = {
    schemas: {} as Record<string, APISchema>,
  };

  for (const r of Object.values(api.resources)) {
    const schema = r.schema;
    const [collection, parentPWPS] = generateParentPatternsWithParams(r);

    if (parentPWPS.length === 0) {
      parentPWPS.push({ pattern: "", params: [] });
    }

    const patterns: string[] = [];
    const schemaRef = `#/components/schemas/${r.singular}`;
    const singular = r.singular;

    const bodyParam = {
      required: true,
      content: {
        "application/json": {
          schema: { $ref: schemaRef },
        },
      },
    };

    const idParam = {
      in: "path",
      name: singular,
      required: true,
      schema: { type: "string" },
    };

    const resourceResponse = {
      description: "Successful response",
      content: {
        "application/json": {
          schema: { $ref: schemaRef },
        },
      },
    };

    for (const pwp of parentPWPS) {
      const resourcePath = `${pwp.pattern}${collection}/{${singular}}`;
      patterns.push(resourcePath.slice(1));

      if (r.listMethod) {
        const listPath = `${pwp.pattern}${collection}`;
        const responseProperties: Record<string, APISchema> = {
          results: {
            type: "array",
            items: { $ref: schemaRef },
          },
          nextPageToken: {
            type: "string",
          },
        };

        if (r.listMethod.hasUnreachableResources) {
          responseProperties.unreachable = {
            type: "array",
            items: { type: "string" },
          };
        }

        const params = [
          ...pwp.params,
          {
            in: "query",
            name: "maxPageSize",
            required: false,
            schema: { type: "integer" },
          },
          {
            in: "query",
            name: "pageToken",
            required: false,
            schema: { type: "string" },
          },
        ];

        if (r.listMethod.supportsSkip) {
          params.push({
            in: "query",
            name: "skip",
            required: false,
            schema: { type: "integer" },
          });
        }

        if (r.listMethod.supportsFilter) {
          params.push({
            in: "query",
            name: "filter",
            required: false,
            schema: { type: "string" },
          });
        }

        addMethodToPath(paths, listPath, "get", {
          operationId: `List${kebabToPascalCase(singular)}`,
          description: `List method for ${singular}`,
          parameters: params,
          responses: {
            "200": {
              description: "Successful response",
              content: {
                "application/json": {
                  schema: {
                    type: "object",
                    properties: responseProperties,
                  },
                },
              },
            },
          },
        });
      }

      if (r.createMethod) {
        const createPath = `${pwp.pattern}${collection}`;
        const params = [...pwp.params];
        if (r.createMethod.supportsUserSettableCreate) {
          params.push({
            in: "query",
            name: "id",
            required: false,
            schema: { type: "string" },
          });
        }

        addMethodToPath(paths, createPath, "post", {
          operationId: `Create${kebabToPascalCase(singular)}`,
          description: `Create method for ${singular}`,
          parameters: params,
          requestBody: bodyParam,
          responses: {
            "200": resourceResponse,
          },
        });
      }

      if (r.getMethod) {
        addMethodToPath(paths, resourcePath, "get", {
          operationId: `Get${kebabToPascalCase(singular)}`,
          description: `Get method for ${singular}`,
          parameters: [...pwp.params, idParam],
          responses: {
            "200": resourceResponse,
          },
        });
      }

      if (r.updateMethod) {
        addMethodToPath(paths, resourcePath, "patch", {
          operationId: `Update${kebabToPascalCase(singular)}`,
          description: `Update method for ${singular}`,
          parameters: [...pwp.params, idParam],
          requestBody: {
            required: true,
            content: {
              "application/merge-patch+json": {
                schema: { $ref: schemaRef },
              },
            },
          },
          responses: {
            "200": {
              description: "Successful response",
              content: {
                "application/merge-patch+json": {
                  schema: { $ref: schemaRef },
                },
              },
            },
          },
        });
      }

      if (r.deleteMethod) {
        const params = [...pwp.params, idParam];
        if (r.children.length > 0) {
          params.push({
            in: "query",
            name: "force",
            required: false,
            schema: { type: "boolean" },
          });
        }

        addMethodToPath(paths, resourcePath, "delete", {
          operationId: `Delete${kebabToPascalCase(singular)}`,
          description: `Delete method for ${singular}`,
          parameters: params,
          responses: {
            "204": {
              description: "Successful response",
              content: {
                "application/json": {
                  schema: {},
                },
              },
            },
          },
        });
      }

      for (const custom of r.customMethods) {
        const methodType = custom.method.toLowerCase();
        const cmPath = `${resourcePath}:${custom.name}`;
        const methodInfo = {
          operationId: `:${kebabToPascalCase(custom.name)}${kebabToPascalCase(
            singular
          )}`,
          description: `Custom method ${custom.name} for ${singular}`,
          parameters: [...pwp.params, idParam],
          requestBody: {},
          responses: {
            "200": {
              description: "Successful response",
              content: {
                "application/json": {
                  schema: custom.response,
                },
              },
            },
          },
        };

        if (custom.method == "POST") {
          methodInfo.requestBody = {
            required: true,
            content: {
              "application/json": {
                schema: custom.request,
              },
            },
          };
        }

        addMethodToPath(paths, cmPath, methodType, methodInfo);
      }
    }

    const parents = r.parents.map((p) => p.singular);
    schema.xAEPResource = {
      singular: r.singular,
      plural: r.plural,
      patterns,
      parents,
    };

    components.schemas[r.singular] = schema;
  }

  for (const [k, v] of Object.entries(api.schemas)) {
    components.schemas[k] = v;
  }

  const contact = api.contact
    ? {
        name: api.contact.name,
        email: api.contact.email,
        url: api.contact.url,
      }
    : undefined;

  return {
    openapi: "3.1.0",
    servers: [{ url: api.serverURL }],
    info: {
      title: api.name,
      version: "version not set",
      description: `An API for ${api.name}`,
      contact,
    },
    paths,
    components,
  };
}

export function generateParentPatternsWithParams(
  resource: Resource
): [string, PathWithParams[]] {
  if (resource.patternElems.length > 0) {
    const collection = `/${
      resource.patternElems[resource.patternElems.length - 2]
    }`;
    const params: any[] = [];
    for (let i = 0; i < resource.patternElems.length - 2; i += 2) {
      const pElem = resource.patternElems[i + 1];
      params.push({
        in: "path",
        name: pElem.slice(1, -1),
        required: true,
        schema: { type: "string" },
      });
    }

    const pattern = resource.patternElems
      .slice(0, resource.patternElems.length - 2)
      .join("/");
    return [
      collection,
      [
        {
          pattern: pattern ? `/${pattern}` : "",
          params,
        },
      ],
    ];
  }

  const collection = `/${collectionName(resource)}`;
  const pwps: PathWithParams[] = [];

  for (const parent of resource.parents) {
    const singular = parent.singular;
    const basePattern = `/${collectionName(parent)}/{${singular}}`;
    const baseParam = {
      in: "path",
      name: singular,
      required: true,
      schema: { type: "string" },
      xAEPResourceRef: { resource: singular },
    };

    if (parent.parents.length === 0) {
      pwps.push({
        pattern: basePattern,
        params: [baseParam],
      });
    } else {
      const [_, parentPWPS] = generateParentPatternsWithParams(parent);
      for (const parentPWP of parentPWPS) {
        const params = [...parentPWP.params, baseParam];
        const pattern = `${parentPWP.pattern}${basePattern}`;
        pwps.push({ pattern, params });
      }
    }
  }

  return [collection, pwps];
}

function addMethodToPath(
  paths: Record<string, OpenAPIPathItem>,
  path: string,
  method: string,
  methodInfo: any
): void {
  let methods = paths[path];
  if (!methods) {
    methods = {};
    paths[path] = methods;
  }

  switch (method) {
    case "get":
      methods.get = methodInfo;
      break;
    case "post":
      methods.post = methodInfo;
      break;
    case "patch":
      methods.patch = methodInfo;
      break;
    case "put":
      methods.put = methodInfo;
      break;
    case "delete":
      methods.delete = methodInfo;
      break;
  }
}

export interface PathItem {
  get?: {
    parameters?: Array<{
      in: string;
      name: string;
      required: boolean;
      schema: APISchema;
      xAEPResourceRef?: {
        resource: string;
      };
    }>;
    responses: Record<
      string,
      {
        description: string;
        content?: {
          "application/json": {
            schema: APISchema;
          };
        };
      }
    >;
  };
  post?: {
    parameters?: Array<{
      in: string;
      name: string;
      required: boolean;
      schema: APISchema;
      xAEPResourceRef?: {
        resource: string;
      };
    }>;
    requestBody?: {
      required: boolean;
      content: {
        "application/json": {
          schema: APISchema;
        };
      };
    };
    responses: Record<
      string,
      {
        description: string;
        content?: {
          "application/json": {
            schema: APISchema;
          };
        };
      }
    >;
  };
  patch?: {
    parameters?: Array<{
      in: string;
      name: string;
      required: boolean;
      schema: APISchema;
      xAEPResourceRef?: {
        resource: string;
      };
    }>;
    requestBody?: {
      required: boolean;
      content: {
        "application/json": {
          schema: APISchema;
        };
      };
    };
    responses: Record<
      string,
      {
        description: string;
        content?: {
          "application/json": {
            schema: APISchema;
          };
        };
      }
    >;
  };
  delete?: {
    parameters?: Array<{
      in: string;
      name: string;
      required: boolean;
      schema: APISchema;
      xAEPResourceRef?: {
        resource: string;
      };
    }>;
    responses: Record<
      string,
      {
        description: string;
        content?: {
          "application/json": {
            schema: APISchema;
          };
        };
      }
    >;
  };
}

export function parseOpenAPI(openapi: OpenAPI): OpenAPI {
  return openapi;
}

export function getPathItem(
  openapi: OpenAPI,
  path: string
): PathItem | undefined {
  return openapi.paths[path];
}

export function getSchema(schema: APISchema | undefined): APISchema {
  if (!schema) {
    return { type: "object", properties: {} };
  }
  if ("$ref" in schema && schema.$ref) {
    const refSchema: APISchema = {
      type: "object",
      properties: {},
      $ref: schema.$ref,
    };
    return refSchema;
  }
  return {
    ...schema,
    type: schema.type || "object",
  };
}

export function getMethod(
  openapi: OpenAPI,
  path: string,
  method: string
): PathItem[keyof PathItem] | undefined {
  const pathItem = getPathItem(openapi, path);
  if (!pathItem) {
    return undefined;
  }
  return pathItem[method as keyof PathItem];
}

export function getParameters(
  openapi: OpenAPI,
  path: string,
  method: string
):
  | Array<{
      in: string;
      name: string;
      required: boolean;
      schema: APISchema;
      xAEPResourceRef?: {
        resource: string;
      };
    }>
  | undefined {
  const methodItem = getMethod(openapi, path, method);
  if (!methodItem || !("parameters" in methodItem)) {
    return undefined;
  }
  return methodItem.parameters;
}

export function getRequestBody(
  methodInfo: PathItem[keyof PathItem]
): RequestBody | undefined {
  if (!methodInfo || !("requestBody" in methodInfo)) {
    return undefined;
  }
  const requestBody = (methodInfo as { requestBody?: RequestBody }).requestBody;
  return requestBody;
}

export function getResponses(
  openapi: OpenAPI,
  path: string,
  method: string
):
  | Record<
      string,
      {
        description: string;
        content?: {
          "application/json": {
            schema: APISchema;
          };
        };
      }
    >
  | undefined {
  const methodItem = getMethod(openapi, path, method);
  if (!methodItem || !("responses" in methodItem)) {
    return undefined;
  }
  return methodItem.responses;
}
